Database implementation to store data

CastagnaIT opened the pull request:
## Types of changes
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] I have read the [**CONTRIBUTING**](Contributing.md) document.

The basic idea is to implement a database management.
This should allow us to solve all the concurrency and data loss problems, between service and frontend addon, and should allow us to better manage different parts of code and remove some workaround.

To first i have already transferred the locale_id to db, seem to work but need other tests

The next steps in my mind are:
✔save profiles to db witch allow to know the last used profile directly
✔save session data to db
✔store some hidded data from xml to db like password and username
✔store temporany menu data to db
✔store library data to db
✔store stream continuity data to db [by profile]


**Fixed issues**
Locale id error - https://github.com/CastagnaIT/plugin.video.netflix/issues/78
Automatic library updates - https://github.com/CastagnaIT/plugin.video.netflix/issues/74 - https://github.com/CastagnaIT/plugin.video.netflix/issues/126
Solved concurrency and data loss problems of persistent_storage and some settings
Fixed logout now the profiles are no longer accessible and you can enter your credentials - https://github.com/CastagnaIT/plugin.video.netflix/issues/25
Fixed exporting tvshow.nfo for a single episode
Fixed UpNext from kodi library watched status work
Fixed issue with library items that containing % in the path
Fixed continuous "new access" email notification from netflix - https://github.com/CastagnaIT/plugin.video.netflix/issues/93

**Improvements**
Implemented a version upgrade system
Implemented management of login errors, now you should have more details of the access problems

**Features introduced in relation to database**
Implemented automatic export of new episodes
Added a new context menu "Export new episodes" for a single tv show of library to perform the operation manually
Ability to exclude (and re-include) a tv show from library auto update
Possibility to share the same library database with multiple devices that using netflix addon, in order to work it is necessary have a mysql server, you can follow the official kodi MySQL setup instructions at https://kodi.wiki/view/MySQL, then set the connection parameters also in the addon settings.

a little at a time... beer please

### All Submissions:

* [X] Have you followed the guidelines in our Contributing document?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?


### Changes to Core Features:

* [X] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you successfully ran tests with your changes locally?

## Screenshots (if appropriate):

[You can erase any parts of this template not applicable to your Pull Request.]

CastagnaIT commented:
@smeulf check and comment my progress here

Smeulf commented:
A comment about stream continuity.

That would be awesome if we can share the data between multiple clients, and have it 'per profile'.

Many times, I like to watch TV shows in English with subtitles. But my wife prefers French, for the same tv show she's watching on the bedroom ^^

But if I move to the bedroom, then I can't retrieve my own preferences...

So I wouldn't add the stream continuity to the db, but more in a separated sharable db.

For the rest, I'll take a look by the week and run some tests :)

CastagnaIT commented:
no need to make more db, simply divide the information by profile with an db relation
you can explore db with sqlitestudio

Smeulf commented:
That works for a single client, but won't for multiple clients at the same time, mostly if you store the last profile Id in the db. Or you'll have to store the client associated...

CastagnaIT commented:
multiple client in what sense?

Smeulf commented:
One Kodi in the living room, one in the bedroom, one on a tablet, etc...

My setup is that way, and Kodi db is running on MariaDB.

The idea is to share stream continuity data between all of them, through SMB or NFS.

Then if I change the language somewhere, for my profile, I can get the changes everywhere.

So it requires a separate db, or it willing be to complex to manage.

CastagnaIT commented:
now i understand, so yours is a particular thing so interesting,
but have another database to manage requires work to effect inter-relationships from two database manually via code, things that should be automatic from db system
i don't find this solution very acceptable, maintain two databases and create one database only for a single smallest list is pretty ridiculous
I'm going to simplify things and leave the operations for cascade changes to the database system

so what could be done?
-you can create your custom mod addon
-dont touch stream_continuity, but i prefer to move it to handle data automatically
-other solution?

CastagnaIT commented:
it is not will overwritten, the code is dedicated to copy only if there isn't
so as i wrote in the comment if the creation management from scratch is not too complicated, i expect to implement it
to now i need go faster current db is not definitive

Smeulf commented:
> it is not will overwritten

Oh ok, it's a template, not in the definitive folder, copied to the right place if not present :) Sorry for miss understanding that :)

About the 'extra db', I do understand your point.

I was also thinking that at the end, it could contains the library db also. Why? Well for some compatibility with the stream workaround that queries the database to work. And if the library database is not shared, it's not working on my particular setup.

I propose we go step by step:
- Let's finish the database implementation first
- Move the stream continuity to the db, with multiple profile handling

=> We will have (more than) the same level of functionality than today, and maybe it can be enough (setting the language once for all is not so hard, and it may not worth the work to have it automatic)

- Then I can write some code to move the stream continuity again to a different db **with** the library, and submit a PR to provide that functionality to everyone if you like it.

I don't really like having to move it again, but that sounds the easiest way for now. Plus if it fails, we will have all in the db already.

CastagnaIT commented:
well let's see the thing later for now i already have several changes to make
yes we can move also the library file to db, many things can be done😊

Smeulf commented:
I wonder if the database upgrade should not be better at service startup. Then the check would be performed only once, not on every addon call. But you may have some own reasons to make it that way :)

Also, can you please add those fields in the AppConfig table? So I can use them on my stuff as soon as you pushed that work. Thanks.

export_all_new_episodes_running (integer 0/1)
export_all_new_episodes_start_time (text for datetime)
library_auto_update_last_start (text for date)

CastagnaIT commented:
-yes upgrade to the database i can move it to service

-only the change of version is checked at each start, but the code for the changes is not executed if the version does not change, not seem to me that can do better than this

-to get or store your data you can use: eg:
g.DB.get_value('export_all_new_episodes_running', 0)
g.DB.set_value('export_all_new_episodes_running', 1)
all data is stored will converted as string, i do not use different types

but i suggest you wait to make implementations the thing is not ready,
i have to make other changes

but are you on vacation that you never stop!?

Smeulf commented:
> but are you on vacation that you never stop!?

Kinda ^^ But not for long hopefully!

Smeulf commented:
> only the change of version is checked at each start

Yes but it requires database access that may not be unnecessary, mostly on some low speed storages used on RPi ;). That's not a lot, but all time that can be saved should be ;) It's already pretty long to process the all stuff on a Pi so the less we add the better it is.

CastagnaIT commented:
in any case the access to database must be made, it is not those two variables that make the difference...also because the db connections are reused not lost

you lucky.. i don't know yet if they allow me to stay home at least 5 days...
I would like to stay at home for at least 2 days... :(

Smeulf commented:
As always, I'm sharing my own opinions, and you remain the boss!

Take care with the heat!

Smeulf commented on resources/lib/database.py:
```diff
@@ -0,0 +1,257 @@
+    def _get_active_guid_profile(self):
+        query = 'SELECT Guid FROM Profiles WHERE IsActive = 1'
+        cur = self._execute_query(query)
+        return cur.fetchone()[0]
```
With the template database just copied, I have an error at this line, and can't list my profiles.

See log https://paste.kodi.tv/puvuzanebu.kodi

CastagnaIT replied:
no profiles appears to have been saved to db, it is strange
did you delete my personal profile data from the db? being no longer related it can cause problems

Smeulf replied:
No profile on the db, no credential asked.

The db is the last you pushed, with no profile inside.
I also tried to removed my settings.xml to enfoce credential asking, without success.

If it helps, here's a full debug log when I totally remove the userdata\addon_data\plugin.video.netflix folder.
https://paste.kodi.tv/oyoheyapex.kodi

Smeulf replied:
And a full debug log after upgrading (before I delete the folder
https://paste.kodi.tv/xekubosujo.kodi

CastagnaIT replied:
the main problem seem that the Session table has disappeared
SQLite error no such table: Session

Smeulf replied:
There's also this after the session is initialised in the "upgrade" log

Login prefetch: No stored credentials are available

Smeulf commented:
Can you add an AutoUpdate table in the shared DB? I'd like to store the export-new-episodes data in the shared db as it will be library related.

# Shared DB Table mapping: { Table name : [ columns ]}
TABLE_AUTO_UPDATE = {'AutoUpdate': ['Name', 'Value']}

I'll add the library table myself later when I'll work on it.

CastagnaIT commented:
find a better table name, this is too general

Smeulf commented:
What about 'SharedAppConfig' then? So it can save more data at the end :)

CastagnaIT commented:
good👌

Smeulf commented on resources/lib/globals.py:
```diff
@@ -215,6 +215,17 @@ def init_globals(self, argv):
+        import resources.lib.database.db_shared as db_shared
+        import resources.lib.database.db_utils as db_utils
+        # TODO: xml settings to specify a custom path
+        #  need to study how better apply to client/service
```
For the client, it's not a real problem IMO as it will be checked at every lanch.

For the service, what I did for the library updater is to make it a xbmc.Monitor instance, and compute the time again at event 'onSettingsChanged'. Maybe it can be done the same way for all services?

Or we need to send an AddonSignal when settings_monitor detects a setting change.

Smeulf commented on resources/lib/globals.py:
```diff
@@ -215,6 +215,17 @@ def init_globals(self, argv):
+        import resources.lib.database.db_shared as db_shared
+        import resources.lib.database.db_utils as db_utils
+        # TODO: xml settings to specify a custom path
+        #  need to study how better apply to client/service
```
For the client, it's not a problem as it's checked on every launch.

For the service, I'd use the settings_monitor and 'onSettingChange' compare the current DB path to the settings DB path.

**If changed:**
If the file already exists in the new path, use it (it's most probably a file moved from an another client)
Else if it doesn't exists, move (more copy and delete) the current db file to the new location
change self.SHARED_DB to the new value

Smeulf replied:
We also need to ensure the path is avaliable, ie the SMB folder is not offline, to not create a file again in addons_data/plugin.video.netflix

CastagnaIT replied:
I can say that this is what i thought, you added some more details
but now i have two other things that i have to change, so then i'll think about it
we will see next week

Smeulf commented:
No more database template?

Smeulf commented:
Thanks. Premiminary tests are now very promissing! No show stopper! Good game!

CastagnaIT commented:
I think we misunderstood the other day...
I think you wanted to do more or less the things i wanted,
but only that i intended to completely eliminate videoid (the class value) from database
the code you see is NOT finished, and NOT tested -WIP-
I preferred to divide the categories of data into different tables, more understandable and easier to maintain

I just need to understand one thing, why do you need to search an episode or movie starting from the file name?

Smeulf commented:
> but only that i intended to completely eliminate videoid (the class value) from database

That was my goal too, indeed we didn't understand each other :)

> I just need to understand one thing, why do you need to search an episode or movie starting from the file name?

Not sure it's absolutely needed, the way you did it looks good to me.

Also we must not forgot this commit https://github.com/Smeulf/plugin.video.netflix/commit/d2a722b3a61bd0a1b7edeadfd9fa91c45f07f09a (must be adapted).

Do you want to keep working on the library or do you want me to?

Smeulf commented:
/!\ In the lib/kodi/library.py and lib/services/llibrary_updater.py, some g.PERSISTENT_STORAGE remains. They should be moved to the shared_db app_config table.

Smeulf commented:
Fast test says the 'is_in_library' function is not working as expected: it returns true for at least a tvshow or season not in the library, and so the context menus are not the good ones. For an empty library I have 'refresh' and 'delete from library' context menus items.

Browsing episodes is failing because of an SQL error for now: https://paste.kodi.tv/ahusikajam.kodi

Yes it's WIP, but the faster we know, the faster we can fix it ;)

CastagnaIT commented:
I will give a look but in this moment it is not yet ready to run :(
...also in db missing relations

so later i finish some parts, when I'm done i tell you
surely you have more time than me to test my changes and make fixes
 
I also have two possible improvements to make you do, but let's see later...

Smeulf commented:
Just pointed the most obvious bugs ;)
Keep me posted as soon as I can be helpful!

Also, I'd save the translated path in the db, instead of the special path if any, using 'xbmc.translatepath'. Then we don't need to translate the paths anywhere after this. But that's an improvement I can do and test later.

Eventually,  we will need an upgrade function from the old 'library.db2' file to the sqlite3 db, that should be something I can write.

CastagnaIT commented:
I think i have fixed all the main bugs
i prefer save special path and leave the task of path resolution to kodi
now also library_updater.py should be ok

so now need to do some test to check is there is some other bug
in another day i do the refactor of stream continuity data

there is also todo:
1) migration code library.db2 to shared_db
2) library.py: purge() it can be redone using only the database, without calling the netflix server to get metadata that is not needed
3) a question: exclude_from_update can it be self-configuring based on elapsed time?
for example: after 1 month without updates to a TV show, automatically exclude it from updates
It might be a good idea to limit checking for unnecessary updates and lighten the server load

Smeulf commented:
Much better now :)

I have just a problem with unicode chars I think: SQLite error You must not use 8-bit bytestrings unless you use a text_factory that can interpret 8-bit bytestrings. I added manifest data in the debug log.

https://paste.kodi.tv/iraperilas.kodi ('é' chars in the path)
https://paste.kodi.tv/omekugutab.kodi (Æ char in the path)

For points 1 & 2, I'll handle them tomorrow hopefully, and you'll be able to cherry pick the commits.

For point 3, that's a damn good question, with a very complicated answer. We have multiple situations with NF.
a - New episodes every week
b - New episodes once a year (or 2 years, see The OA), a full season at a specific date
c - New acquiried seasons, that can show up anytime
d - A show that is considered ended that will have a new season 2 years after (ie Money Heist)

Plus, there's nothing in the manifest that shows upcomming episodes (see Cable Girls metadata from this log https://paste.kodi.tv/iraperilas.kodi) even if a new season is planned (August 9th)

So unless we can find a way to query Netflix upcomming page, that would mean we disable the auto update forever until turned on again. Then we must remember the date it was turned on, for a month, etc. Not very user friendly IMO :(

CastagnaIT commented:
hi,
fixed the reported bugs

Cable Girls metadata link is wrong :(

there is a Netflix upcomming page in website? or only to android/ios app?
if so, you should try to track the requests/answers that are made by the app
at the moment i have no idea if it's possible

CastagnaIT commented:
I just tried monitoring traffic of netflix app on android with a proxy
but seem the communications used is with ssl and encrypted with a certificate
at least this way it doesn't seem possible to get some info to get upcoming data

Smeulf commented:
I have to report my work on point 1 & 2. Will work on it during the week.

About upcoming, I'll take a look too, but without great hope...

Sorry about the delay.

Smeulf commented:
> library.py: purge() it can be redone using only the database, without calling the netflix server to get metadata that is not needed

The more I think about it, the more I feel it's the 'remove' method that needs a rework, to avoid using metadata, not the 'purge' itself...

This will end with a very new method to build tasks for removal, and maybe a rename of current build tasks methods that will remain for adding items only.

CastagnaIT commented:
> I would put them in the shared db instead, but both can work.
> 
> We must recommend our users to setup the auto update only on one client when the shared db will be movable.

I can implement a verify based on a specific computer a sort of "uuid"
so does understand to the addon from which computer run the auto-update
but i still have to think about it, i believe it is necessary to add a new setting, to allow user set the "current pc" to allow auto-update
a sort of "Set this pc to perform auto-update", when set automatically exclude others shared devices

Smeulf commented:
That's a good idea for sure. I was thinking the same, but couldn't find a suitable way to implement it.

That means it needs a uuid in the local database of every client, and a field to store the allowed client in the shared db, then we compare both.

The main problem will be with settings: how to show that the actual client is the active one? A disabled boolean maybe, activated by a button to make the client active and set the db value. And then, show the update options only when the client is active...

CastagnaIT commented:
The uuid is not a problem i already have the solution
so for the settings options, i'll think about it
for now i suspend this implementation for next week,
I'm completing other things

Smeulf commented:
Tried to enable auto export new episodes, got this error: https://paste.kodi.tv/ocaxeluroj.kodi

Maybe I can fix it tomorrow, but not certain.

Smeulf commented on resources/lib/upgrade_controller.py:
```diff
@@ -0,0 +1,53 @@
+
+def check_addon_upgrade():
+    """Check addon upgrade and perform necessary update operations"""
+    addon_previous_ver = g.LOCAL_DB.get_value('addon_previous_version', g.VERSION)
```
If there's no 'addon_previous_version' value in the db, then previous version = g.VERSION (so the current version)

Smeulf commented on resources/lib/upgrade_controller.py:
```diff
@@ -0,0 +1,53 @@
+    """Check addon upgrade and perform necessary update operations"""
+    addon_previous_ver = g.LOCAL_DB.get_value('addon_previous_version', g.VERSION)
+    addon_current_ver = g.VERSION
+    if addon_current_ver != addon_previous_ver:
```
as 'addon_previous_ver' was set to current version, they are never different, so 'addon_previous_ver' is never set in the DB, so update won't ever occur.

addon_previous_ver should be set to something in the default DB imo :)

CastagnaIT replied:
whooops 😓 i made a loop
fortunately you help me to verify

Smeulf replied:
I can fix it with 'None' as default value in the commit I have to make for library migration. I wish to use the add-on controller to perform that operation.

Smeulf replied:
Plus I'd like to import username and password from the xml file if some exists, to avoid entering credentials again, if you agree :)

CastagnaIT replied:
i have already included fix to the original commit

> Plus I'd like to import username and password from the xml file if some exists, to avoid entering credentials again, if you agree :)

aaah no is just a passage, it would become dead code within a release...
someone will be annoyed but will pass😂

Smeulf replied:
As you wish boss :D

Smeulf replied:
> i have already included fix to the original commit

What's the default value you used?

Smeulf commented:
Or ’addon_previous_ver = g.LOCAL_DB.get_value('addon_previous_version', g.VERSION)’ should have an another default value than g.VERSION

Smeulf commented:
Looks like you momved the ESN to the database. But, what if a user wants to change it manually (Android 4K)? :/

CastagnaIT commented:
Today i did some tests to add the menu to set the custom shared-database folder

The main problem is that settings_monitor is in the service thread,
so to update the g.SHARED_DB in the addon thread,
the more practical and clean solution would be "common.send_signal" to send a signal to the addon thread to run a func to update g.SHARED_DB with the new path.

The problem is that in my tests the signal is sent, but it does not arrive at its destination to start the callback function.
I have not yet figured out what the problem is..

CastagnaIT commented:
*Implemented version upgrade controller
changed, a little more fat but more configurable in case of need

Smeulf commented:
About the path for the shared library: you don't need to inform the add-on thread with a Signal.

Just call the add-on with a specific route, send a Signal to the service, the service will update the db paths.

Once the add-on is started again, it will read the correct data.

Addon thread is life limited as an action (display something, start something, etc) and ends at the end of that action.

Only the service thread is a loop with unlimited lifetime...

Smeulf commented:
Or just implement it in settings monitor, not even need of a signal...

CastagnaIT commented:
Thanks tomorrow do a try

Smeulf commented:
You can take a look in the library_updater at onSettingChanged if you need some inspiration :)

CastagnaIT commented:
see if there is something to fix in the last commit

for the device selection for auto-update, a question 
why can't auto update be done from any device?
so the last check is saved in the database

Smeulf commented:
> why can't auto update be done from any device?

I see two possible problems if performed from different devices:
- Netflix server overload leading to ban
- A possible database locking issue if both devices updates the all episodes at the same time 

Plus for me, it makes no sense to export strm files and update a shared Kodi library with multiple devices at the same time. As it's shared, once updated by one device, all the other can see the new records.

Nevertheless, the manual 'export new episodes' from a menu item should remain accessible by any device.

CastagnaIT commented:
good but all device they must be able to export
otherwise to example a person to add a single movie is obliged to turn on a second device only to export the movie to the library and then turn it off ... it would make little sense

Smeulf commented:
Except my previous comment, the code looks good. As it seems we're getting very close of the end of the enhancement, I'll install it on my 2 RPi, and run some extended tests. For... let say a week before merging?

Smeulf commented:
> good but all device they must be able to export
otherwise to example a person to add a single movie is obliged to turn on a second device only to export 

Yes. Only the auto update must be restricted to one single device.

CastagnaIT commented:
> Except my previous comment, the code looks good. As it seems we're getting very close of the end of the enhancement, I'll install it on my 2 RPi, and run some extended tests. For... let say a week before merging?

yes no problem
if possible before the next weekend, because i take advantage of it because i am at home

I see to implement this last thing

Smeulf commented:
> if possible before the next weekend

I'm deploying right now, and as my wife is on holidays, she won't have any other option than test it lol

Smeulf commented:
For safety, can you please change https://github.com/CastagnaIT/plugin.video.netflix/blob/e935b04e195f0ce4a45c8215afba4b4a742e7406/resources/lib/upgrade_actions.py#L37 `if episode_key != 'videoid':` by `if episode_key not in ['videoid', 'nfo_export']:`

In a previous version, that key was present. It won't worth a lot to exclude it in case someone did already try. Thanks.

CastagnaIT commented:
yes i add it

Smeulf commented on resources/lib/kodi/context_menu.py:
```diff
@@ -27,15 +27,15 @@ def ctx_url_builder(videoid):
         'label': common.get_local_string(30196),
-        'url': ctx_item_url(['excludefromautoupdate'], g.MODE_LIBRARY)},
-    'includeinautoupdate': {
+        'url': ctx_item_url(['exclude_from_autou_pdate'], g.MODE_LIBRARY)},
```
Small typo leading to an error ^^ 
Should be exclude_from_auto_update

CastagnaIT replied:
🚀

Smeulf commented:
And an exception when trying to play a video, I think related to the last ESN change: https://paste.kodi.tv/efuzegocov.kodi

CastagnaIT commented:
you used custom or automatic esn?

Smeulf commented:
It was automatically set in the settings.xml by the previous version, so I don't know what to answer, I think you consider it as manual in this case right?

CastagnaIT commented:
yes in this case the current behaviour, if there is esn in xml, it is not set to db
having the db blank, the esn is not set so the issue of None value

CastagnaIT commented:
before there was no distinction between manual and automatic
an possibility solution is, reset esn in xml in addon upgrade and advise the user to update it if he had set it manually

Smeulf commented:
I'll update again on next push :) Then I'll test the shared db.

Smeulf commented:
Why not just build an auto ESN in the db if None? And then use the one form the xml to override the one from db?

For a previous setup, it will always override, but does it matter so much?

Smeulf commented:
A note about stream_continuity behabior:

If I set a TV show to English audio, and French subtitles, stream_continuity will correctly remember the settings for my profile.

Now, my wife wants to watch the same tv show, but in french, no subs (default settings). She starts the first episode. We may expect as there's no data in the database, it will use the defaults settings for Kodi (French). 

BUT, as there is a but... As the episode was already played once, Kodi remembers the streams used, and the episode will start in English with subs.

She will have to change the settings again in her profile for stream_continuity to work.

This behavior will change once we include the profile id in the video link: the files won't be the same for Kodi anymore, so no previous play settings will apply.

Hope I'm clear enough :D

CastagnaIT commented:
ah now understand because before her you have tried watching it in English
how boring this mix of settings and override... prrr

Smeulf commented:
Seems like we have a problem to open a database moved to a shared folder, like "smb://server/share/"

If I read well, as it's an URI, it should use a specific URI form. See here at sqlite3.connect(https://docs.python.org/3/library/sqlite3.html) and here(https://www.sqlite.org/uri.html)

Smeulf commented:
Plus not even sure it will work :/

Smeulf commented:
At worse, I can mount the remote share...

CastagnaIT commented:
here there is some uri example https://www.sqlite.org/c3ref/open.html#urifilenameexamples
the main problem is the file lock
multiple connections can read simultalely but only one can write
Seem there is some solutions:
1) Write-Ahead Logging or WAL
2) wal2
about halfway down the page there is some short info:
https://stackoverflow.com/questions/4060772/sqlite-concurrent-access
I have not studied in depth, we need to understand how it works

Smeulf commented:
Not an URI as I expected... So for Linux OS, it will require a mount point. Good to know.

I start wondering if we wouldn't save time by implementing a MySQL interface if the shared db is moved...

CastagnaIT commented:
would be what kodi does, use sqlite for local databases and mysql for shared
so requires a lot work,
need to rewrite all the shared part, split the files, make a create database/table, check database existant and connection...
moreover it is better to divide the names of the files for the base classes of access to the database with more explicit names based on the type of database ex. db_base_sqlite and db_base_mysql
and add a dedicated menu for the settings to all configurations of cases
if we take this way the release of this update will unfortunately be delayed...

CastagnaIT commented:
so sql is more stable and safe of sqlite,
i've read the wal yes is a choice, but have some unpleasant situation mainly: 
-depend from VFS, that VFS depend by the type of operating system so we cannot take care to test all operating systems...
-if for some reason a connection is not closed, a temporary WAL file remains on the disk with the operation not completed, the guide does not specify what happens after this
if we want to keep sharing

I was forgetting something else too, there is also have to do a sort of interface to excchange the connection from local sqlite shared-database to shared-mysql and transfer the existing data!?

everything is possible but i think the thing is getting complex especially for maintenance

Smeulf commented:
Well, I imagine it 'simple': when using a MySQL db, just create the new db, ensure profiles are in the db, and that's all. No data gathering from the previous setup.

When using MySQL in Kodi, it's usually installed at setup. If someone is capable to manage MySQL database, I assume he can import the data from the previous db himself if he wants to.

For the technical implementation, looks like only the connection must be changed.

So either db_base must be splitted in 2 or 3 classes: db_base_sqlite, db_base_mysql and db_base if some functions are common, either db_base can handle both type of connection (my prefered way if we change the error handling messages).

Then we can build the shared_db based on the settings. If MySQL server address user and pass are set, we use MySQL connector, else SQLite3.

If I'm correct, the rest of the code can remain the same (?), except the database creation that should be fully done by code instead of copying db files.

Also maybe the requests should be made ANSI 92 compatible (use GROUP BY when there's an agregate - count, min, etc... - in the select). No need to rewrite the all code.

Smeulf commented:
> Luckily python MySQL Connector module is included in kodi

I checked before summiting the idea :) hope it works on Android too...

Smeulf commented:
I forgot about db selection with MySQL ^^

CastagnaIT commented:
I will take a look at the operation

CastagnaIT commented:
A firt try here: https://github.com/CastagnaIT/plugin.video.netflix/tree/wip_database_mysql
Premise:
-mainly i tried to separate the modules (sqlite/mysql) with their objects as much as possible, except for fetchall/fetchone, if necessary, it can easily be done in the future, but I have doubts that those methods may change their functioning in the future.
-developed only the classes, without ever starting addon
-some parts are yet missing/incomplete
Problems Todo:
-shared database class, the decorator not sure it works
-the incorporated module of kodi "myconnpy" it is very old and DEPRECATED, do not recommend the use of this version, i will make an attempt to incorporate the updated module

Smeulf commented:
> the incorporated module of kodi "myconnpy" it is very old and DEPRECATED, do not recommend the use of this version, i will make an attempt to incorporate the updated module

Not sure the update it will be compatible with LibreELEC, must be tested.

I gonna do some tests on my side to see if the decorator works (have to create the db first).

As a note for later, the MySQL port must be configurable in settings ;)

CastagnaIT commented:
I just wanted to show you what I'm doing, so it is not yet operational, it will not work if you run the addon
I will send you the DDL to create the database exact parameters must be set to create shema and tables

Smeulf commented:
> so it is not yet operational

Yep got it don't worry, I can see how far it is ;) I know it will crash somewhere.

Smeulf commented:
Reading the code, I don't understand why you didn't kept get_cursor, get_value and so on in the base class... Certainly something I'm missing ;)

CastagnaIT commented:
yes I didn't include it because in the base interface i only kept the base functions to make the database to work, while get_value and so on are methods like others to get/set the database data

Smeulf commented:
Ok I wouldn't have choose an interface but a real base class, but both way can work, plus an interface can be extended to other db type, so it's most probably the right choice :)

CastagnaIT commented:
now addon starts, attached the ddl of mysql db i use
I lost a lot of time for the decorators solution, now i found an acceptable solution...
so i didn't do any tests...only checked if addon starts
[TMP_Netflix_DDL_0.1.sql.zip](https://github.com/CastagnaIT/plugin.video.netflix/files/3498633/TMP_Netflix_DDL_0.1.sql.zip)

CastagnaIT commented:
as i thought there is sqlite code to run differently on mysql
at least for all the parts in which "update or insert" is executed
it is necessary to create two different sql codes

I would just like to know if the addon mysql works also from a different device (so not pc)
can you give it a try?
I warn you that for the problem described above it starts correctly only the first time, then you have to empty the tables, but the log tells you if it connect
if it works i continue the coding

Smeulf commented:
I can't create videolibtvshows on LibreELEC (MariaDB official addon) beacuse of an ERROR 1273 (HY000): Unknown collation: 'utf8mb4_0900_ai_ci'

Shouldn't we use the same as Kodi?
```
ENGINE=InnoDB DEFAULT CHARSET=utf8;
/*!40101 SET character_set_client = @saved_cs_client */;
```

CastagnaIT commented:
there is an error in ddl I did not see
utf8mb4 is supported is the current new utf8 encoding
the "utf8" is the old encoding, that real name is utf8mb3

CastagnaIT commented:
voila
[TMP_Netflix_DDL_0.2.sql.zip](https://github.com/CastagnaIT/plugin.video.netflix/files/3501198/TMP_Netflix_DDL_0.2.sql.zip)

Smeulf commented:
I've got a SQLite error. Looks like it doesn't even try to connect to MySQL (MariaDB).

I tried to rename the old 'nf_shared.sqlite3' to 'nf_shared.sqlite3.bak' too, no change.

https://paste.kodi.tv/uqelifovid.kodi

CastagnaIT commented:
have you updated your branch with last modify?

Smeulf commented:
Downloaded the branch from GitHub as zip 2 hours ago before posting the error at db creation.

CastagnaIT commented:
try to force inherit to:
class NFSharedDatabase(db_base_mysql.MySQLDatabase):

Smeulf commented:
> try to force inherit to:
class NFSharedDatabase(db_base_mysql.MySQLDatabase):

This way it works. 

Just a big warning: here the **tables names are case sensitive**, and returns  MySQL error 1146 meaning "table doesn't exist"

CastagnaIT commented:
strange to me works without problem at least in official MySQL server
mariadb its a fork... why you use mariadb?

CastagnaIT commented:
does the connection work anyway? can I continue or worry?
you can temporarily rename the tables to test

Smeulf commented:
For a test, I renamed the 'profile' table to 'Profiles'. The service starts now, and the table contains the expected data.

So yes, it's connecting from RPi. Do you have any way to try on Android? 
Anyway, it's connecting on the main platforms.

Smeulf commented:
And why MariaDB? 
- Because it's the official addon from LibreELEC.
- It's also MariaDB that Synology provides as database, and I know synology is used by some to host Kodi DB.

CastagnaIT commented:
very well, mmh i have only a honor 8 not very suitable for testing but i can give a try

CastagnaIT commented:
tested addon works database connected and store data

Smeulf commented:
Btw, for now the wip_database brahch works fine! No error to report until now.

CastagnaIT commented:
i have noticed now that in your librelec with kodi 18.1 there is sqlite ver 3.26
in my kodi 19 for windows 3.17 i didn't understand why they don't have the same versions

CastagnaIT commented:
I need you to do a test on mariaDB
to first, try to re-create the database with this ddl
I need to know if the ddl script is compatible with mariadb

[TMP_Netflix_DDL_0.3.sql.zip](https://github.com/CastagnaIT/plugin.video.netflix/files/3505725/TMP_Netflix_DDL_0.3.sql.zip)

I believe I have corrected the major mysql addon bugs
when you can do a test

ps. i have renamed all table names, sqlite files are updated

CastagnaIT commented:
I think we need to add a menu in the settings,
to allow the deletion of the mysql database, accessible even if you use sqlite shareddb (in case of fallback or problems)

Smeulf commented:
We're getting good, mostly.

ddl script is compatible, I recreated the DB using it, no error.

Using SQLite DB for the shared db is still working.

I have an error when exporting a tv show to db: `MySQL error 1062 (23000): Duplicate entry '80074220' for key 'PRIMARY'` (80074220 is the show id)
https://paste.kodi.tv/nodigusoqo.kodi

Smeulf commented:
New log with debug and extra debug info:
https://paste.kodi.tv/ekuvojulin.kodi

The error is in insert_season

Smeulf commented:
Forgot to mention, the error is with MySQL only.

CastagnaIT commented:
I can't reproduce the error in my mysql server, it works...
or mariadb not support some type of query or is other

CastagnaIT commented:
you can try to execute manually with a software ui this query, anche check is the result is correct 0 or 1 if exist
'SELECT EXISTS('
                 'SELECT 1 FROM video_lib_seasons '
                 'INNER JOIN video_lib_tvshows '
                 'ON video_lib_seasons.TvShowID = video_lib_tvshows.TvShowID '
                 'WHERE video_lib_tvshows.TvShowID = ? AND video_lib_seasons.SeasonID = ?)'

Smeulf commented:
`SELECT EXISTS(
SELECT 1 FROM video_lib_seasons
INNER JOIN video_lib_tvshows
ON video_lib_seasons.TvShowID = video_lib_tvshows.TvShowID
WHERE video_lib_tvshows.TvShowID = 80074220 AND video_lib_seasons.SeasonID = 80074825)` returned 1

CastagnaIT commented:
if you change with non existing ID's return 0?

Smeulf commented:
video_lib_seasons.SeasonID = 8007482**4** returned 0 yes, as not present in the db

CastagnaIT commented:
that strange... 
you can put in "insert_season" some debug output to verify what type of value is returned from self.season_id_exists() suggest str(...) and also str(type(...))

CastagnaIT commented:
so not best inside season_id_exists method

Smeulf commented:
I found the issue!

TvShowID is the only primary key for the seasons table. SeasonID primary key is missing ^^

```
CREATE TABLE netflix_addon.video_lib_seasons (
  TvShowID INT(11) NOT NULL,
  SeasonID INT(11) NOT NULL,
  PRIMARY KEY (TvShowID)
)
ENGINE = INNODB,
CHARACTER SET utf8mb4,
COLLATE utf8mb4_unicode_ci;
```

CastagnaIT commented:
just change the database then╰(*°▽°*)╯

Smeulf commented:
That's what I did lol :D 
You just need to update the ddl on your side. 

Tomorrow, I'll run tests from multiple clients, and check stream_continuity and resume workaround are working. We're getting good for a release soon. (I'll have to do the french translations before releasing)

About your question for dropping the shared db, I'd say 'no'.

If they change from SQL to local, they are good enough to clean the DB by themself. Anyone working with MySQL should have sufficient knowledge to manage it IMO. It's not a configuration a 'standard user' will use.

And if the addon can't connect to the shared db, we should pop-up an error message, not a notification, and people have to fix the issue. No fallback, the end. A but brutal, but good enough to me.

CastagnaIT commented:
you are a devil hahaha😂 so ok

Smeulf commented:
I'm just being pragmatic lol.

Should it be a real problem, I promise I'll write the code myself ;)

Smeulf commented:
A note on MariaDB table names: the case sensitive name depends on the OS. On Windows it's not, on Linux it is. Voilà.

Smeulf commented:
I would have generate the unique Id once at database init, based on time, computer name and maybe some other parameters to make it unique, and then store that value in the local db. So when you upgrade the hardware or restore data, there nothing to configure again.

I fear the uuid based on Mac address causes issue if you change ie just from Ethernet to WLAN.

CastagnaIT commented:
computer date or name is not good, a fixed parameter is required and the mac address is used in many commercial realities
for now this code can remain so, it is what is also used to encrypt the credentials
to be sure you always have the same mac address you just need to understand how to get the complete mac list and sort it alphabetically, is a fix that can be done later

Smeulf commented:
So be it!

CastagnaIT commented:
if you find other defects / problems let me know, then we start the release
the translations are still to be done

Smeulf commented:
I couldn't test everything yet, sorry was out for the all day. Hopefully I will be able to run some tests in the evening.

About the release, regarding how big is the update, should it be beta first? Should it upgrade major version? Like 0.15.0-beta1 ?

For the languages, I'll update the french PR with new strings so you'll just have to merge it.

Smeulf commented:
I got an error while importing an old version of the library.db, I have to check and fix it.
https://paste.kodi.tv/afemociwes.kodi

Smeulf commented:
Also a new error: `KeyError: u'episodes'` while fetching episodes list.

https://paste.kodi.tv/samahipeta.kodi

Smeulf commented:
You're saving a lot of data in the MySQL db for stream continuity. The all history? Is'nt it too much?

![image](https://user-images.githubusercontent.com/1092948/63201504-cfbd8c00-c085-11e9-871f-30b5c1746536.png)

Smeulf commented:
I think the table creation is once again wrong.

```
CREATE TABLE netflix_addon.stream_continuity (
  ID INT(11) NOT NULL AUTO_INCREMENT,
  ProfileGuid VARCHAR(50) NOT NULL,
  VideoID INT(11) NOT NULL,
  Value TEXT DEFAULT NULL,
  DateLastModified VARCHAR(50) NOT NULL,
  PRIMARY KEY (ID)
)
ENGINE = INNODB,
CHARACTER SET utf8mb4,
COLLATE utf8mb4_unicode_ci;
```

should be (ID column dropped, primary keys changed)

```
CREATE TABLE netflix_addon.stream_continuity (
  ProfileGuid VARCHAR(50) NOT NULL,
  VideoID INT(11) NOT NULL,
  Value TEXT DEFAULT NULL,
  DateLastModified VARCHAR(50) NOT NULL,
  PRIMARY KEY (ProfileGuid, VideoID)
)
ENGINE = INNODB,
CHARACTER SET utf8mb4,
COLLATE utf8mb4_unicode_ci;
```

CastagnaIT commented:
> I couldn't test everything yet, sorry was out for the all day. Hopefully I will be able to run some tests in the evening.
> 
> About the release, regarding how big is the update, should it be beta first? Should it upgrade major version? Like 0.15.0-beta1 ?
> 
> For the languages, I'll update the french PR with new strings so you'll just have to merge it.

yes agreed major update bump, but not beta
this is not the case with long-term multi-stage development, there will be only a few bugs to fix



> Also a new error: `KeyError: u'episodes'` while fetching episodes list.
> 
> https://paste.kodi.tv/samahipeta.kodi

I don't understand what you were doing, more details


> I think the table creation is once again wrong.
> 
> ```
> CREATE TABLE netflix_addon.stream_continuity (
>   ID INT(11) NOT NULL AUTO_INCREMENT,
>   ProfileGuid VARCHAR(50) NOT NULL,
>   VideoID INT(11) NOT NULL,
>   Value TEXT DEFAULT NULL,
>   DateLastModified VARCHAR(50) NOT NULL,
>   PRIMARY KEY (ID)
> )
> ENGINE = INNODB,
> CHARACTER SET utf8mb4,
> COLLATE utf8mb4_unicode_ci;
> ```
> 
> should be (ID column dropped, primary keys changed)
> 
> ```
> CREATE TABLE netflix_addon.stream_continuity (
>   ProfileGuid VARCHAR(50) NOT NULL,
>   VideoID INT(11) NOT NULL,
>   Value TEXT DEFAULT NULL,
>   DateLastModified VARCHAR(50) NOT NULL,
>   PRIMARY KEY (ProfileGuid, VideoID)
> )
> ENGINE = INNODB,
> CHARACTER SET utf8mb4,
> COLLATE utf8mb4_unicode_ci;
> ```

I suspected then i forgot... you are right

Smeulf commented:
>> Also a new error: KeyError: u'episodes' while fetching episodes list.

>> https://paste.kodi.tv/samahipeta.kodi

> I don't understand what you were doing, more details

I was just trying to enter into the episodes listing for a season. I think it's the same type of error as the 'lists' error you fixed a couple of days ago. A random error I couldn't see before. Not related to this specific dev, we can switch it as an issue.

Smeulf commented:
About stream continuity, the VideoID saved in the DB is the episode one, not the id of the show.
I think this can also explain why I had so much entries.

CastagnaIT commented:
sorry but i don't have time now tomorrow morning i have to go celebration person who is no longer on earth... if you fix something you know how to do it
then tomorrow afternoon i should have some time

Smeulf commented:
I'm sorry for you and the familly. May that person rest in peace.
You'll fix it when you can.

Smeulf commented:
We may have a show stopper: when I switched from local shared db to MySQL shared db on my second RPi, it seems the service didn't reload the shared db. 
Then I had a lot of error until I restart kodi.

I have to check again and provide debug logs, hopefully tomorrow.

Smeulf commented:
> I got an error while importing an old version of the library.db, I have to check and fix it.
https://paste.kodi.tv/afemociwes.kodi

i think the commit https://github.com/CastagnaIT/plugin.video.netflix/commit/581fcb6b4087d94ae7bcbe2ef3c8df45fcc654bd broke the compatibility with very old files.

If you take a look at the log, it's trying to reteive `self._id_values[5]` for a tvshowid, and can't find it. The commit moved the tvshowid from_id_values[4] to _id_values[5], and then, this very old VideoID from a very old db file can't be imported.

The only thing we can do is to catch any error in migrate_library_to_db and finally rename the old file.
You're opinion is welcome...

Smeulf commented:
Alternate way:

I can list the files names in the db, then open the files, and build a videoid from the path in the file, if the file still exists. Else if the file doesn't exists, I don't import the video.

Smeulf commented on resources/language/resource.language.en_gb/strings.po:
```diff
@@ -811,3 +811,35 @@ msgstr ""
+msgid "Username"
+msgstr ""
+
+msgctxt "#30202"
```
Duplicated string, same as #30004

CastagnaIT replied:
little pain, it is not worth wasting time to change all the commits and the code for a string
in case if necessary we can reuse it for something else

Smeulf commented on resources/language/resource.language.en_gb/strings.po:
```diff
@@ -811,3 +811,35 @@ msgstr ""
+msgstr ""
+
+msgctxt "#30206"
+msgid "ERROR: The MySQL database is not reachable - check the connection - fallback to local database"
```
Are we really falling back?

CastagnaIT replied:
yes if mysql not work instantiates local mysql sqlite database,
when the connection is corrected, resume with mysql
if you prefer to improve the error phrase, propose one

CastagnaIT replied:
now that I think better, there is a problem
the addon restarts when an action occurs so the connection when working is automatically restored, but in the service remains the old one once initializated ....mmmh....

CastagnaIT replied:
I think the easiest way is that in case of connection problems we disable via code the setting
"use_mysql" to False and so fallback to sqlite.
this way when the user tries to reactivate the mysql connection via settings
it is also updated in the service (after the change to be made)

CastagnaIT replied:
I try to do something then when you have time you let me know if it could be fine

CastagnaIT replied:
I forgot global is already reinitialized when settings change so g.SHARED_DB shoud be yet changed... now I am confused i wait your rpi log

Smeulf commented:
Another annoying issue: I miss the 'add to my list' context menu entry, still can't understand why.

Seems `g.LOCAL_DB.get_profile_config('isKids', False)` returns the right value, but also seems the expression `if videoid.mediatype in [common.VideoId.MOVIE, common.VideoId.SHOW] and not g.LOCAL_DB.get_profile_config('isKids', False):` in context_menu.py doesnt work.

CastagnaIT commented:
> Another annoying issue: I miss the 'add to my list' context menu entry, still can't understand why.
> 
> Seems `g.LOCAL_DB.get_profile_config('isKids', False)` returns the right value, but also seems the expression `if videoid.mediatype in [common.VideoId.MOVIE, common.VideoId.SHOW] and not g.LOCAL_DB.get_profile_config('isKids', False):` in context_menu.py doesnt work.

fixed

CastagnaIT commented:
later i will watch your fixes

CastagnaIT commented:
> > I got an error while importing an old version of the library.db, I have to check and fix it.
> 
> https://paste.kodi.tv/afemociwes.kodi
> 
> i think the commit [581fcb6](https://github.com/CastagnaIT/plugin.video.netflix/commit/581fcb6b4087d94ae7bcbe2ef3c8df45fcc654bd) broke the compatibility with very old files.
> 
> If you take a look at the log, it's trying to reteive `self._id_values[5]` for a tvshowid, and can't find it. The commit moved the tvshowid from_id_values[4] to _id_values[5], and then, this very old VideoID from a very old db file can't be imported.
> 
> The only thing we can do is to catch any error in migrate_library_to_db and finally rename the old file.
> You're opinion is welcome...

> I got an error while importing an old version of the library.db, I have to check and fix it.

> https://paste.kodi.tv/afemociwes.kodi
> 
> i think the commit https://github.com/CastagnaIT/plugin.video.netflix/commit/581fcb6b4087d94ae7bcbe2ef3c8df45fcc654bd broke the compatibility with very old files.
> 
> If you take a look at the log, it's trying to reteive `self._id_values[5]` for a tvshowid, and can't find it. The commit moved the tvshowid from_id_values[4] to _id_values[5], and then, this very old VideoID from a very old db file can't be imported.
> 
> The only thing we can do is to catch any error in migrate_library_to_db and finally rename the old file.
> You're opinion is welcome...

I think I can handle the problem by adding compatibility for old objects, later I try to create a fix for you to test

CastagnaIT commented:
I added the migration fix with old videoid, need you to do a test

> We may have a show stopper: when I switched from local shared db to MySQL shared db on my second RPi, it seems the service didn't reload the shared db.
> Then I had a lot of error until I restart kodi.
> 
> I have to check again and provide debug logs, hopefully tomorrow.

I think the service remains in operation permanently even when the addon restarts
so will need to add in SettingsMonitor a check for mysql settings change
and update var in global

Smeulf commented:
> I added the migration fix with old videoid, need you to do a test

Tested and fix proposal in last PR

> better understand better what happens

I'll setup a test MySQL server in the night, and post logs.

Smeulf commented:
Well I thought I could reproduce the issue under Windows but I can't.

I'll have to try again under LE on a specific SD Card, I don't have the logs anymore. Anyway, after a reboot it works on my second RPi so it's not something so bad.

CastagnaIT commented:
i have changed some strings and reused the double one

Smeulf commented:
> i have changed some strings and reused the double one

Language PR updated

Smeulf commented:
Please just pick the French commit, didn't rebased my branch before updating.

Smeulf commented:
> We may have a show stopper: when I switched from local shared db to MySQL shared db on my second RPi, it seems the service didn't reload the shared db.
> Then I had a lot of error until I restart kodi.
> 
> I have to check again and provide debug logs, hopefully tomorrow.

Well I double checked, and everything is ok.
The issue was because the service was not running, because it crashed when importing the old library (with old videoids). Now everything updated smoothly. Even export new episodes seems good with MySQL so seems all lights are now green.

Smeulf commented:
A new error minor: https://paste.kodi.tv/deyuzijuhe.kodi

Sorry again it's not in debug, catched by my wife :) (tks to her!) She just tried to enter the episodes listing from a season. 

Next atteempt, it worked. Should I catch it in debug I'll post a new log.

CastagnaIT commented:
> A new error minor: https://paste.kodi.tv/deyuzijuhe.kodi
> 
> Sorry again it's not in debug, catched by my wife :) (tks to her!) She just tried to enter the episodes listing from a season.
> 
> Next attempt, it worked. Should I catch it in debug I'll post a new log.

i suggest also add a debug output to path_response in EpisodeList()
it is strange that in the answer lacks id season
maybe a particular case? remember what tvshow?

Smeulf commented:
I can retrieve the show name, but it was a 1 time error. On the next attempt it worked fine, so no way to get the debug logs as I can't reproduce it.

What could be interesting would be to output of the API call result on the error channel in case of an error in shakti. I suspect it's some kind of network error where the result is empty...

CastagnaIT commented:
if it is an error that does not happen often we can fix it later,
I wait until tomorrow evening and then release

Smeulf commented:
Have to say you did an amazing job! Thank you!

CastagnaIT commented:
I thank you, you did your part too
many things are solved thanks to your help

dm2912 commented:
dont know if its related, but 0.15.1 is not working for me. since update i get a sqlite error and then it refuses to load anything.

specific line in log reads

sqliteerror: no such table app_config 

(might have app-config) but anyway,,.

Smeulf commented:
@dm2912 can you provide full debug log please?

dm2912 commented:
I've managed to sort it for now.

I basically uninstalled completely and removed entire userdata folder.

reinstalled, let it do its thing, and then restored my library folders and just the nf_shared database (which seems to be the one with the library entries in)

so its running a clean vanilla main db and it seems to be sorted

if it comes back will do a full log. 

it was insisting no such table but i grabbed the sql and looked at it and the table was most definitely there so possibly my db just got corrupted or summat.

Smeulf commented:
Ok for now then.
I'm just a bit worried as you're not the olny one with this error :/
I doubled checked the database tamplate in the afternoon, and the app_config table is there...
There's must be something we missed, but what?! 
And for now, we don't have any debug log with this error, only normal logging, so no way to understand what's going on.

dm2912 commented:
yeh, i immediately grabbed a copy of the sql file and had a look at it and the table was definitely there. 
[nf_local.sqlite3.zip](https://github.com/CastagnaIT/plugin.video.netflix/files/3548497/nf_local.sqlite3.zip)

fwiw attached my original db file (username/password removed), the table definitely there

Smeulf commented:
I do belive you even without the db ;)
Let's wait to catch it in debug, or with a full code inspection, but I don't know where to loock at :(

dm2912 commented:
yeh, it was really just in case you spotted something awry in my db that would cause an issue..

turned on kodi this morning, immediate sqlite error.. 

[kodi.log](https://github.com/CastagnaIT/plugin.video.netflix/files/3551687/kodi.log)

error at line 3849

dr88dr88 commented:
I had the same error, I deactivated the export to library schedule and the problem was solved.

dm2912 commented:
> I had the same error, I deactivated the export to library schedule and the problem was solved.

Did you disable both:

Keep my List and Kodi Library in Sync
and
Export New Episodes

or just the latter?

Smeulf commented:
Didn't look at logs yet, bit if it's related to the auto update, I suspect it's trying to request the wrong table in the shared-db, app_data instead of shared_app_data.

I'll take a look in the evening. Thanks for your reports guys!

Smeulf commented:
@dm2912 would you try this branch before I sumbit a PR? https://github.com/Smeulf/plugin.video.netflix/tree/Fix_shared_app_config_queries

I forced the table name for the queries so it should fix the issue. As I know you"re using export nex episodes, I'd be glad if you report once installed, and in a few days to ensure everything is now working.

dm2912 commented:
installed and re-enabled library updates. will see how it behaves

dm2912 commented:
hmmm, but now library items arent playing...

Smeulf commented:
PR https://github.com/CastagnaIT/plugin.video.netflix/pull/175 updated
